core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

```diff
@@ -791,36 +849,32 @@ public boolean onMicrosoftLoginComplete(PendingMicrosoftAuthentication.Authentic
             return false;
         }
         task.cleanup(); // player is online -> remove pending authentication immediately
-        Throwable ex = task.getLoginException();
-        if (ex != null) {
-            geyser.getLogger().error("Failed to log in with Microsoft code!", ex);
-            disconnect(ex.toString());
-        } else {
-            MsaAuthenticationService service = task.getMsaAuthenticationService();
-            GameProfile selectedProfile = service.getSelectedProfile();
-            if (selectedProfile == null) {
-                disconnect(GeyserLocale.getPlayerLocaleString(
-                        "geyser.network.remote.invalid_account",
-                        clientData.getLanguageCode()
-                ));
-            } else {
-                this.protocol = new MinecraftProtocol(
-                        selectedProfile,
-                        service.getAccessToken()
-                );
-                try {
-                    connectDownstream();
-                } catch (Throwable t) {
-                    t.printStackTrace();
-                    return false;
-                }
+        return task.getAuthentication().handle((result, ex) -> {
+             if (ex != null) {
+                 geyser.getLogger().error("Failed to log in with Microsoft code!", ex);
+                 disconnect(ex.toString());
+                 return false;
+             }
 
-                // Save our refresh token for later use
-                geyser.saveRefreshToken(bedrockUsername(), service.getRefreshToken());
-                return true;
-            }
-        }
-        return false;
+             StepMCProfile.MCProfile mcProfile = result.session().getMcProfile();
+             StepMCToken.MCToken mcToken = mcProfile.getMcToken();
+
+             this.protocol = new MinecraftProtocol(
+                     new GameProfile(mcProfile.getId(), mcProfile.getName()),
+                     mcToken.getAccessToken()
+             );
+
+             try {
+                 connectDownstream();
+             } catch (Throwable t) {
+                 t.printStackTrace();
+                 return false;
+             }
+
+             // Save our auth chain for later use
+             geyser.saveAuthChain(bedrockUsername(), GSON.toJson(result.step().toJson(result.session())));
+             return true;
+         }).join();
```

Comment:
If I understand the code correctly, this `join()` will always be on a done task, *probably*, but I worry this will trip us up in the future. Can we find a way around it?

Reply:
Done